core/trino-main/src/main/java/io/trino/sql/planner/DomainTranslator.java

```diff
@@ -588,6 +603,115 @@ private Map<NodeRef<Expression>, Type> analyzeExpression(Expression expression)
         }
 
+        private Optional<ExtractionResult> createVarcharCastToDateComparisonExtractionResult(
+                ComparisonExpression node,
+                VarcharType sourceType,
+                NullableValue value,
+                boolean complement)
+        {
+            Cast castExpression = (Cast) node.getLeft();
+            Expression sourceExpression = castExpression.getExpression();
+            ComparisonExpression.Operator comparisonOperator = node.getOperator();
+            requireNonNull(value, "value is null");
+
+            if (complement || value.isNull()) {
+                return Optional.empty();
+            }
+            if (!(sourceExpression instanceof SymbolReference)) {
+                // Calculation is not useful
+                return Optional.empty();
+            }
+            Symbol sourceSymbol = Symbol.from(sourceExpression);
+
+            if (!sourceType.isUnbounded() && sourceType.getBoundedLength() < 10) {
+                // too short
+                return Optional.empty();
+            }
+
+            LocalDate date = LocalDate.ofEpochDay(((long) value.getValue()));
+            if (date.getYear() < 1001 || date.getYear() > 9998) {
+                // Edge cases. 1-year margin so that we can go to next/prev year for < or > comparisons
+                return Optional.empty();
+            }
+
+            // superset of possible values, for the "normal case"
+            ValueSet valueSet;
+            boolean nullAllowed = false;
+
+            switch (comparisonOperator) {
+                case EQUAL:
+                    valueSet = dateStringRanges(date, sourceType);
+                    break;
+                case NOT_EQUAL:
+                case IS_DISTINCT_FROM:
+                    if (date.getDayOfMonth() < 10) {
+                        // TODO: possible to handle but cumbersome
+                        return Optional.empty();
+                    }
+                    valueSet = ValueSet.all(sourceType).subtract(dateStringRanges(date, sourceType));
+                    nullAllowed = (comparisonOperator == IS_DISTINCT_FROM);
+                    break;
+                case LESS_THAN:
+                case LESS_THAN_OR_EQUAL:
+                    valueSet = ValueSet.ofRanges(Range.lessThan(sourceType, utf8Slice(Integer.toString(date.getYear() + 1))));
+                    break;
+                case GREATER_THAN:
+                case GREATER_THAN_OR_EQUAL:
+                    valueSet = ValueSet.ofRanges(Range.greaterThan(sourceType, utf8Slice(Integer.toString(date.getYear() - 1))));
+                    break;
+                default:
+                    return Optional.empty();
+            }
+
+            // Date representations starting with whitespace, sign or leading zeroes.
+            valueSet = valueSet.union(ValueSet.ofRanges(
+                    Range.lessThan(sourceType, utf8Slice("1")),
```

Comment:
I understand the purpose of `< 1` , but I don't understand the purpose of `>9`.
Could you please add an explanation for this?

Reply:
a date can start with any whitespace: space, CR, LF, tab, etc.
Whitespace characters are not guaranteed to all go before digits (i don't think they do).